walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty array list as 
+        // the meetingTimes collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration == 0) {
+            return Arrays.asList();
+            }
+
+        // Find avaliable TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, or there is no required attendee, 
+        // return meetingTimes for all attendees.
+        if (allMeetingTimes.size() > 0 || requiredAttendees.size() == 0) {
+            return allMeetingTimes;
+        }
+        // If not, return avaliable TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * Two pointers rangeStart and rangeEnd are pointing at the current avaliable 
+      * TimeRange. If the current avaliable TimeRange has longer duration than the 
+      * request meeting duration, add it to avaliable meetingTimes.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.defaultComparator);
+
+        // The TimeRange of the last event that is went through.
+        TimeRange lastRange = TimeRange.fromStartDuration(DAY_END, 0);
+        int rangeStart = DAY_START; // Pointer to start of the previous event.
+        int rangeEnd = DAY_END; // Pointer to the end of the previous event.
+
+        // Linear search the current events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if the pervious event and current event overlap.
+            if (rangeStart > eventStart) {
```

Comment:
Btw, TimeRange.java actually has an overlaps method!

Reply:
Thanks for the notice, but I think I'll just keep the pointer for clarity! :)